docs/design/node-selection.md

```diff
@@ -0,0 +1,87 @@
+>
+> On the Storj network, preferential storage node reputation is only used to select where new data will be stored, both during repair and during the upload of new files, unlike disqualifying events.  If a storage node’s preferential reputation decreases, its file pieces will not be moved or repaired to other nodes.
+
+The existing reputation-like system uses uptime and audit responses.  It does not currently consider geographic location, throughput, or latency.  In addition to factors which affect reputation, there are other factors in node selection.  These considerations currently include IP address, advertised available bandwidth, advertised available disk space, software version compatibility, and whether the node appeared to be online in the latest communication with the satellite.
+
+One final factor involved in node selection is node 'vetting.'  During upload
+
+
+## Design
+
+Separate reputation scores will be maintained for audits and uptime.  Two sets of configurations will be stored, two different reputation scores will be calculated, etc..  
+
+The reputation _R(n)_ is calculated based on some measured success _v_, two persisted values a and β, and fixed configuration values λ and _w_.  
+
+> α(n) = λ·α(n-1) + _w_(1+_v_)/2
+>
+> β(n) = λ·β(n-1) + _w_(1-_v_)/2
+>
+> R(n) = α(n) / (α(n) + β(n))
+
+Initial values for α and β - α0 and β0 - will also be configuration values.  While _v_ may need to vary depending on our "easing" implementation, it will likely vary algorithmically and should not require configuration.  The initial implement may simply assume that _v_ = 1 on success and _v_ = -1 on failure.
+
+Existing codes which updates database audit / uptime success-counts and ratios must be updated to track audit / uptime α and β values.  `TotalAuditCount` and `TotalUptimeCount` will still be needed to determine if the node is new or vetted.
+
+The node selection SQL queries will also change.  Twice as many nodes must returned from these functions to satisfy the "Power of Two Choices" requirement, which gives preference to nodes with better reputation.  For every two nodes returned, the one with the higher reputation scores will be selected returned and the other discarded.
+
+Note that the initial implementation has two different reputation statistics:  audit and uptime.  For the purposes of node selection, we assume that these two reputation can be combined by scaling one of them by some constant.  We further assume that different operations may weigh these reputations differently. For instance, repair may be more concerned that a node is reliable than it is speedy.  New file uploads coming from an uplink may have different criteria.   The initial configuration should include a `uptime_repair_weight`, `audit_repair_weight`, `uptime_uplink_weight`, and `audit_uplink_weight` constants.
+
+> Total Repair Reputation = uptime_repair_weight · uptime R(n) + audit_repair_weight · audit R(n)
```

Comment:
I'm thinking in the implementation of these totals and how `uptime R(n)` and `audit R(n)` are going to be correlated, because each check runs in a different process but we need both results.

My understanding is that `n` is the number of the check which if they were executed at the same moment `n` will be value of the counter (henceforth `N`), however the checks aren't coordinated so we could store the last one and then combined when the both `n` are registered, but it won't work because the frequency of each check could be (or it may already be) different, so the one with lower frequency will never catch the one with higher.

For example:

* Audit runs every 60 seconds
* Uptime runs every 10 seconds.

__After 10 seconds__

* Audit_N = 0
* Uptime_N = 1
	* Uptime(1): OK

__After 20 seconds__

* Audit_N = 0
* Uptime_N = 2
	* Uptime(2): OK

...

__After 60 seconds:__

* Audit_N = 1
	* Audit(1): OK
* Uptime_N = 6
	* Uptime(6): Error

Total Reputation Repair (1) = uptime_repair_weight * uptime R(1) + audit_repair_weight * audit R(1)
Total Uplink Reputation (1) = uptime_uplink_weight * uptime R(1) + audit_uplink_weight * audit R(1)

...

__After 90 seconds:__

* Audit_N = 1
* Uptime_N = 9
	* Uptime(9): OK

...

__After 120 seconds:__

* Audit_N = 1
	* Audit(2): Error
* Uptime_N = 12
	* Uptime(12): OK

Total Reputation Repair (2) = uptime_repair_weight * uptime R(2) + audit_repair_weight * audit R(2)
Total Uplink Reputation (2) = uptime_uplink_weight * uptime R(2) + audit_uplink_weight * audit R(2)


Storing the values for each `n` doesn't look feasible and coordinate both process doesn't seem a good idea.

Reply:
While the `n` in `R(n)` effectively equals `TotalAuditCount` or `TotalUptimeCount`, these values are not used when calculating `R(n)`.  At any given point in time, audit and uptime are expected to be using different effective values of `n`, due to them being measured at different rates.

Reply:
so basically it's

`Total Repair Reputation = uptime_repair_weight · uptime R(n) + audit_repair_weight · audit R(m)`

The question is now how to implement it, because those uptimes and audits run in different Go routines so it could be done in different way, but we can have that definition in the ticket that will tackle such task.